adapters/nanointeractive/nanointeractive.go

```diff
@@ -0,0 +1,199 @@
+package nanointeractive
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"net/http"
+)
+
+type NanointeractiveAdapter struct {
+	endpoint string
+}
+
+// used for cookies and such
+func (a *NanointeractiveAdapter) Name() string {
+	return "Nano"
+}
+
+func (a *NanointeractiveAdapter) SkipNoCookies() bool {
+	return false
+}
+
+func (a *NanointeractiveAdapter) MakeRequests(bidRequest *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	var errs []error
+	var validImps []openrtb.Imp
+
+	var adapterRequests []*adapters.RequestData
+
+	for i := 0; i < len(bidRequest.Imp); i++ {
+		err := checkImp(&bidRequest.Imp[i])
+		// If the parsing is failed, remove imp and add the error.
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		validImps = append(validImps, bidRequest.Imp[i])
+	}
+
+	if len(validImps) == 0 {
+		errs = append(errs, fmt.Errorf("no impressions in the bid request"))
+		return nil, errs
+	}
+
+	// set referrer origin
+	if refO := getRefererOrigin(&bidRequest.Imp[0]); refO != "" {
+		if bidRequest.Site == nil {
+			bidRequest.Site = &openrtb.Site{}
+		}
+		bidRequest.Site.Ref = refO
+	}
+
+	bidRequest.Imp = validImps
+
+	reqJSON, err := json.Marshal(bidRequest)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+	headers.Add("User-Agent", getSafeUa(bidRequest.Device))
+	headers.Add("X-Forwarded-For", getSafeIp(bidRequest.Device))
+	headers.Add("Referer", getSafeReferrer(bidRequest.Site))
```

Comment:
Do we want to set the "User-Agent", "X-Forwarded-For", or the "Referer" header even if `bidRequest.Device` or `bidRequest.Site` are `nil`? Do you think we want to condition setting those headers to `Device` and `Site` not being `nil`? I'm just double checking, maybe your want this headers to always be added. Please let me know.
```
 26   func (a *NanointeractiveAdapter) MakeRequests(bidRequest *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
 27  
 28   *-- 34 lines: var errs []error-------------------------------------------
 62  
 63       headers := http.Header{}
 64       headers.Add("Content-Type", "application/json;charset=utf-8")
 65       headers.Add("Accept", "application/json")
 66       headers.Add("x-openrtb-version", "2.5")
 67 -     headers.Add("User-Agent", getSafeUa(bidRequest.Device))
 68 -     headers.Add("X-Forwarded-For", getSafeIp(bidRequest.Device))
 69 -     headers.Add("Referer", getSafeReferrer(bidRequest.Site))
    +     if bidRequest.Device != nil {
    +         headers.Add("User-Agent", bidRequest.Device.UA)
    +         headers.Add("X-Forwarded-For", bidRequest.Device.IP)
    +     }
    +     if bidRequest.Site != nil {
    +         headers.Add("Referer", bidRequest.Site.Page)
    +     }
 70  
 71   *-- 12 lines: set user's cookie-----------------------------------------
 83       return adapterRequests, errs
 84   }
 85  
 86   *-- 69 lines: func (a *NanointeractiveAdapter) MakeBids(----------------
155   func getSafeIp(device *openrtb.Device) string {
156       if device == nil {
157           return ""
158       }
159       return device.IP
160   }
161  
162   func getSafeUa(device *openrtb.Device) string {
163       if device == nil {
164           return ""
165       }
166       return device.UA
167   }
168  
169   func getSafeReferrer(site *openrtb.Site) string {
170       if site == nil {
171           return ""
172       }
173       return site.Page
174   }
adapters/nanointeractive/nanointeractive.go
```

Reply:
thanks for this, it's more clear like this without unnecessary safe functions